File: src/algebra1_quadratic_formula_medium.py
"""
algebra1 - quadratic_formula (medium)
Generated: 2026-02-22T04:02:09.307734
"""

from problem_utils import *

def _latex_eq(a, b, c):
    """Build readable LaTeX string for ax^2 + bx + c = 0."""
    terms = []
    if a == 1:
        terms.append("x^2")
    elif a == -1:
        terms.append("-x^2")
    else:
        terms.append(f"{a}x^2")
    if b > 0:
        terms.append(f"+ {b}x")
    elif b < 0:
        terms.append(f"- {abs(b)}x")
    if c > 0:
        terms.append(f"+ {c}")
    elif c < 0:
        terms.append(f"- {abs(c)}")
    return " ".join(terms) + " = 0"

def generate():
    target_elo = randint(1300, 1600)

    if target_elo < 1400:
        # a=1, integer roots, wide coefficient range
        a_val = 1
        r1 = randint(-10, 10)
        r2 = randint(-10, 10)
        b_val = -(r1 + r2)
        c_val = r1 * r2
        difficulty = (1300, 1400)

    elif target_elo < 1500:
        if randint(0,1):
            # a∈{2,3,4,5}, integer roots
            a_val = choice([2, 3, 4, 5])
            r1 = randint(-6, 6)
            r2 = randint(-6, 6)
            b_val = -a_val * (r1 + r2)
            c_val = a_val * r1 * r2
        else:
            # a=1, irrational roots from varied b,c
            a_val = 1
            b_val = choice([-8,-6,-4,-2,2,4,6,8])
            c_val = choice([-5,-4,-3,-2,-1,1,2])
        difficulty = (1400, 1500)

    else:
        if randint(0,1):
            # Complex roots: force disc < 0
            a_val = choice([1, 2, 3])
            b_val = choice([-4,-2,2,4,6])
            c_val = choice([5, 8, 10, 13, 17, 20])
            disc_check = b_val**2 - 4*a_val*c_val
            if disc_check >= 0:
                c_val += abs(disc_check)//a_val + 1
        else:
            # a≠1, irrational roots
            a_val = choice([2, 3, 4])
            b_val = choice([-7,-5,-3,3,5,7])
            c_val = choice([-5,-4,-3,-2,2,3])
        difficulty = (1500, 1600)

    disc_val = b_val**2 - 4*a_val*c_val
    roots_syms = [(-b_val + sqrt(disc_val))/(2*a_val),
                  (-b_val - sqrt(disc_val))/(2*a_val)]
    roots_syms_sorted = sorted(roots_syms, key=lambda r: str(r))
    ans = FiniteSet(*roots_syms_sorted)

    eq_str = _latex_eq(a_val, b_val, c_val)

    sol_steps = [
        f"$a={a_val}$, $b={b_val}$, $c={c_val}$",
        f"Discriminant: $\\Delta = ({b_val})^2 - 4({a_val})({c_val}) = {disc_val}$",
    ]
    if disc_val < 0:
        sol_steps.append(f"$x = \\frac{{{-b_val} \\pm {latex(sqrt(disc_val))}}}{{{2*a_val}}}$")
    else:
        sol_steps.append(f"$x = \\frac{{{-b_val} \\pm \\sqrt{{{disc_val}}}}}{{{2*a_val}}}$")
    sol_steps.append(f"Solutions: $x = {latex(roots_syms_sorted[0])}$ and $x = {latex(roots_syms_sorted[1])}$")

    return problem(
        question=f"Solve using the quadratic formula: ${eq_str}$",
        answer=ans,
        difficulty=difficulty,
        topic="algebra1/quadratic_formula",
        solution=steps(*sol_steps),
        calculator="scientific"
    )

emit(generate())

File: src/algebra1_quadratic_formula_medium.jl
# algebra1 - quadratic_formula (medium)
# Generated: 2026-03-08T20:01:04.802778

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x begin
    set_topic!("algebra1/quadratic_formula")
    
    # Choose problem type
    problem_type = choice(1:4)
    
    if problem_type == 1
        # Standard quadratic formula application (1200-1400)
        # ax^2 + bx + c = 0 with integer coefficients but irrational roots
        a = nonzero(-5, 5)
        b = randint(-15, 15)
        c = randint(-20, 20)
        
        # Ensure discriminant is positive but not a perfect square
        disc = b^2 - 4*a*c
        for _attempt in 1:1000
            (disc <= 0 || isqrt(disc)^2 == disc) || break
            b = randint(-15, 15)
            c = randint(-20, 20)
            disc = b^2 - 4*a*c
        end
        
        expr = a*x^2 + b*x + c
        
        # Calculate roots using quadratic formula
        sqrt_disc = sqrt(disc)
        root1 = (-b + sqrt_disc) / (2*a)
        root2 = (-b - sqrt_disc) / (2*a)
        
        # Format answer as set
        ans_str = "\\{\\frac{$(-b) + \\sqrt{$(disc)}}{$(2*a)}, \\frac{$(-b) - \\sqrt{$(disc)}}{$(2*a)}\\}"
        
        problem(
            question="Use the quadratic formula to solve: \$$(tex(expr)) = 0\$",
            answer=ans_str,
            difficulty=(1200, 1400),
            answer_type="expression",
            solution=steps(
                sol("Given", expr ~ 0),
                "Apply quadratic formula: \$x = \\frac{-b \\pm \\sqrt{b^2 - 4ac}}{2a}\$",
                "Here \$a = $(a)\$, \$b = $(b)\$, \$c = $(c)\$",
                "Discriminant: \$\\Delta = $(b)^2 - 4($(a))($(c)) = $(disc)\$",
                sol("Solutions", ans_str)
            ),
            time=120
        )
        
    elseif problem_type == 2
        # Quadratic with one repeated root (1200-1300)
        a = nonzero(-6, 6)
        root = randint(-12, 12)
        
        # Form (x - root)^2, then multiply by a
        expr = expand(a * (x - root)^2)
        
        ans_str = "\\{$(root)\\}"
        
        problem(
            question="Solve using the quadratic formula: \$$(tex(expr)) = 0\$",
            answer=ans_str,
            difficulty=(1200, 1300),
            answer_type="expression",
            solution=steps(
                sol("Given", expr ~ 0),
                "Apply quadratic formula: \$x = \\frac{-b \\pm \\sqrt{b^2 - 4ac}}{2a}\$",
                "Discriminant: \$\\Delta = 0\$ (perfect square)",
                "One repeated root: \$x = $(root)\$",
                sol("Solution", ans_str)
            ),
            time=90
        )
        
    elseif problem_type == 3
        # Quadratic with no real solutions (1300-1500)
        a = nonzero(-4, 4)
        b = randint(-10, 10)
        c = randint(-20, 20)
        
        # Ensure negative discriminant
        disc = b^2 - 4*a*c
        for _attempt in 1:1000
            disc >= 0 || break
            c = randint(-20, 20)
            disc = b^2 - 4*a*c
        end
        
        expr = a*x^2 + b*x + c
        
        problem(
            question="Use the quadratic formula to solve: \$$(tex(expr)) = 0\$. If there are no real solutions, write \"no real solutions\".",
            answer="no real solutions",
            difficulty=(1300, 1500),
            answer_type="text",
            solution=steps(
                sol("Given", expr ~ 0),
                "Apply quadratic formula: \$x = \\frac{-b \\pm \\sqrt{b^2 - 4ac}}{2a}\$",
                "Calculate discriminant: \$\\Delta = $(b)^2 - 4($(a))($(c)) = $(disc)\$",
                "Since \$\\Delta < 0\$, there are no real solutions"
            ),
            time=90
        )
        
    else
        # Word problem requiring quadratic formula setup (1500-1800)
        # Rectangle problem: length is k more than width, area is A
        k = randint(3, 10)
        w = randint(8, 20)
        l = w + k
        area = w * l
        
        # Equation: w(w + k) = area
        # w^2 + kw - area = 0
        
        expr_setup = x^2 + k*x - area
        
        # Discriminant
        disc = k^2 + 4*area
        
        problem(
            question="A rectangle has length $(k) meters more than its width. If the area is $(area) square meters, find the width using the quadratic formula.",
            answer=w,
            difficulty=(1500, 1800),
            solution=steps(
                "Let width = \$x\$ meters, then length = \$x + $(k)\$ meters",
                "Area equation: \$x(x + $(k)) = $(area)\$",
                sol("Simplify", expr_setup ~ 0),
                "Apply quadratic formula: \$x = \\frac{-$(k) \\pm \\sqrt{$(k)^2 + 4($(area))}}{2}\$",
                "Discriminant: \$\\Delta = $(disc)\$",
                "Solutions: \$x = \\frac{-$(k) \\pm $(isqrt(disc))}{2}\$",
                "Taking positive solution: \$x = $(w)\$ meters",
                sol("Width", "$(w) meters")
            ),
            time=180
        )
    end
end